diff --git a/lib/galaxy/tools/data_manager/manager.py b/lib/galaxy/tools/data_manager/manager.py
index 0d6e8ea157d..03befb2eb4b 100644
--- a/lib/galaxy/tools/data_manager/manager.py
+++ b/lib/galaxy/tools/data_manager/manager.py
@@ -326,7 +326,8 @@ def process_result(self, out_data):
             # moving a directory and the target already exists, we move the contents instead
             log.debug('Attempting to add entries for undeclared tables: %s.', ', '.join(data_tables_dict.keys()))
             for ref_file in out_data.values():
-                util.move_merge(ref_file.extra_files_path, self.data_managers.app.config.galaxy_data_manager_data_path)
+                if os.path.exists(ref_file.extra_files_path):
+                    util.move_merge(ref_file.extra_files_path, self.data_managers.app.config.galaxy_data_manager_data_path)
             path_column_names = ['path']
             for data_table_name, data_table_values in data_tables_dict.items():
                 data_table = self.data_managers.app.tool_data_tables.get(data_table_name, None)
@@ -373,7 +374,8 @@ def process_move(self, data_table_name, column_name, source_base_path, relative_
                     if e.errno != errno.EEXIST:
                         raise e
             # moving a directory and the target already exists, we move the contents instead
-            util.move_merge(source, target)
+            if os.path.exists(source):
+                util.move_merge(source, target)
 
             if move_dict.get('relativize_symlinks', False):
                 util.relativize_symlinks(target)
diff --git a/lib/tool_shed/tools/data_table_manager.py b/lib/tool_shed/tools/data_table_manager.py
index 3f1386efabe..e21a8ebf5ad 100644
--- a/lib/tool_shed/tools/data_table_manager.py
+++ b/lib/tool_shed/tools/data_table_manager.py
@@ -169,3 +169,7 @@ def install_tool_data_tables(self, tool_shed_repository, tool_index_sample_files
     def reset_tool_data_tables(self):
         # Reset the tool_data_tables to an empty dictionary.
         self.app.tool_data_tables.data_tables = {}
+
+
+# For backwards compatibility with exisiting data managers
+ToolDataTableManager = ShedToolDataTableManager
